### PR TITLE
Auto resumption

### PR DESCRIPTION
**Pull request by dakinggg**
# What does this PR do?
Add support for autoresume with distributed training.

# What issue(s) does this change relate to?
Fixes [CO-1238](https://mosaicml.atlassian.net/browse/CO-1238?atlOrigin=eyJpIjoiNjdmNTdiZDYzODY3NDAyY2EwOGIyMjczMTc4OTg5ZWEiLCJwIjoiaiJ9)
While fixing this issue, we discovered a race condition that is related to this, that is [CO-1270](https://mosaicml.atlassian.net/browse/CO-1270)

# Manual testing
Manually tested a script that makes a trainer, trains it, makes another trainer with autoresume, and verifies that the params and run name are the same. This was tested on multiples gpus on one node.

Multi node does not currently work, but I also can't really test/debug it because of limited capacity on our cluster.

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/trainer/trainer.py
Comment (hanlint) [inline composer/trainer/trainer.py:1260]: I think we should return `None` here (cc: @mvpatel2000 ?) so that a fresh run with `autoresume: True` still works.

File: composer/trainer/trainer.py
Comment (hanlint) [inline composer/trainer/trainer.py:1190]: nit: `_try_checkpoint_download` ? 

also, perhaps move the `os.makedirs` outside so this function does one thing.

File: composer/trainer/trainer.py
Comment (dakinggg) [reply composer/trainer/trainer.py:1260]: I did have log an info and return `None` at one point. That results in silent failure if the checkpoint doesn't exist/can't be found for some reason. I wasn't sure which was worse, silent failure in the case of it being missing? or erroring on the first run with autoresume true?

Comment (dakinggg) [reply composer/trainer/trainer.py:1260]: I think log a warning and return None makes sense

Comment (hanlint) [reply composer/trainer/trainer.py:1259]: Later on in the codepath, we already log a warning, so i think returning None makes sense.

THe idea here is that the same code/settings while automatically pick up on old checkpoints and resume.

Comment (dakinggg) [reply composer/trainer/trainer.py:1259]: Ah ok, makes sense.

File: composer/trainer/trainer.py
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1250]:
```suggestion
            return latest_checkpoint_path
```
minor nit: if it's raising an error, we don't need to `else` case this. Mostly for consistency with rest of style in codebase

File: composer/trainer/trainer.py
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1260]: same minor nit to unindent this else and remove it

File: composer/trainer/trainer.py
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1285]:
```suggestion
                return latest_checkpoint_path
```
same minor nit

**Review by mvpatel2000 — APPROVED** [review]
Thanks for fixing this!! Approving to unblock, left some really minor nits for conventions we've used in repo. Actually, I'm not sure we're consistent with these... but it'd be nice to do so 🤷

File: composer/trainer/trainer.py
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1230]: Delete since `log.info` prints this later on?

File: composer/trainer/trainer.py
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1242]: remove since we show a progress bar during download?

File: composer/trainer/trainer.py
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1283]: I don't think this barrier is necessary because of the reduce two lines down

File: composer/trainer/trainer.py
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1291]: delete?

**Review by mvpatel2000 — COMMENTED** [review]
I'll let Hanlin approve since it'd be nice to get two pairs of eyes on this. LGTM tho. Amazing job!

File: composer/trainer/trainer.py
Comment (hanlint) [inline composer/trainer/trainer.py:1251]: suggestion: I think both this check and the one below can be unified with `dist.all_gather_object()`, e.g.
```
# list by each rank
latest_checkpoint_exists = dist.all_gather_object(os.path.exists(latest_checkpoint_path))

if self.deepspeed_enabled:
    if not all(latest_checkpoint_exists):
        missing_ranks = [n for (n, exist) in enumerate(latest_checkpoint_exists) if not exist]
        raise RuntimeError(f'Deepspeed was enabled, but checkpoints missing on ranks: {missing_ranks}')
else:
    if not latest_checkpoint_exists[0]:
        return None
```

**Review by hanlint — APPROVED** [review]
LGTM, one suggestion to clean up the code a bit

File: composer/trainer/trainer.py
Comment (dakinggg) [reply composer/trainer/trainer.py:1283]: discussed offline, barrier is necessary so that `os.path.exists` is called after local rank 0 has finished downloading the checkpoint.

File: composer/trainer/trainer.py
Comment (dakinggg) [reply composer/trainer/trainer.py:1242]: I will leave this one because the progress bar doesn't show if the download can't be started for some reason,

File: composer/trainer/trainer.py
Comment (dakinggg) [reply composer/trainer/trainer.py:1291]: This was like the most important line for debugging so I'm going to leave it

File: composer/trainer/trainer.py
Comment (dakinggg) [reply composer/trainer/trainer.py:1251]: yes, this makes sense, thanks

**Comment by dakinggg** [comment]
@hanlint I changed the `num_concurrent_uploads` default to 1. I experimented a bit, and at least on the demo example, if you checkpoint enough to want multiple upload workers, you end up bottlenecked on network communication with s3 (streaming data in and sending checkpoints out). With 1 worker, you should only end up with an invalid symlink if the job dies between finishing uploading the checkpoint and uploading the symlink (which is tiny). This would be quite unlikely, and even this issue could be resolved by having `save_overwrite=True` when using autoresume.

**Review by hanlint — APPROVED** [review]

**Review by eracah — APPROVED** [review]
Approving to unblock!

**Pull request merged**
